openpgp/packet/private_key.go

```diff
@@ -228,95 +253,107 @@ func mod64kHash(d []byte) uint16 {
 	} else {
-		err = pk.serializeUnencrypted(privateKeyBuf)
+		err = pk.serializeUnencrypted(priv)
 	}
-	
 	if err != nil {
 		return
 	}
 
 	ptype := packetTypePrivateKey
-	contents := buf.Bytes()
-	privateKeyBytes := privateKeyBuf.Bytes()
 	if pk.IsSubkey {
 		ptype = packetTypePrivateSubkey
 	}
-	err = serializeHeader(w, ptype, len(contents)+len(privateKeyBytes))
+	err = serializeHeader(w, ptype, contents.Len()+priv.Len())
 	if err != nil {
 		return
 	}
-	_, err = w.Write(contents)
+	_, err = io.Copy(w, contents)
 	if err != nil {
 		return
 	}
-	_, err = w.Write(privateKeyBytes)
+	_, err = io.Copy(w, priv)
 	if err != nil {
 		return
 	}
 	return
 }
 
-func (pk *PrivateKey) serializeDummy(w io.Writer) error {
-	w.Write([]byte{uint8(pk.s2kType)})
-	w.Write([]byte{uint8(pk.cipher)})
-	return pk.s2kParams.Serialize(w)
+func (pk *PrivateKey) serializeDummy(w io.Writer) (err error) {
+	if _, err = w.Write([]byte{uint8(pk.s2kType)}); err != nil {
+		return
+	}
+
+	optional := bytes.NewBuffer([]byte{uint8(pk.cipher)})
+	if err = pk.s2kParams.Serialize(optional); err != nil {
+		return
+	}
+	if pk.Version == 5 {
+		if _, err = w.Write([]byte{uint8(optional.Len())}); err != nil {
+			return
+		}
+	}
+	io.Copy(w, optional)
+	return
 }
 
 func (pk *PrivateKey) serializeEncrypted(w io.Writer) error {
-	privateKeyBuf := bytes.NewBuffer(nil)
-	encodedKeyBuf := bytes.NewBuffer(nil)
-	encodedKeyBuf.Write([]byte{uint8(pk.s2kType)})
-	encodedKeyBuf.Write([]byte{uint8(pk.cipher)})
-	err := pk.s2kParams.Serialize(encodedKeyBuf)
+	encoded := bytes.NewBuffer([]byte{uint8(pk.s2kType)})
+	optional := bytes.NewBuffer([]byte{uint8(pk.cipher)})
+	err := pk.s2kParams.Serialize(optional)
 	if err != nil {
 		return err
 	}
+	optional.Write(pk.iv)
+	if pk.Version == 5 {
+		encoded.Write([]byte{uint8(optional.Len())})
+	}
+	io.Copy(encoded, optional)
 
-	privateKeyBuf.Write(pk.encryptedData)
-
-	encodedKey := encodedKeyBuf.Bytes()
-	privateKeyBytes := privateKeyBuf.Bytes()
-
-	w.Write(encodedKey)
-	w.Write(pk.iv)
-	w.Write(privateKeyBytes)
-
+	if pk.Version == 5 {
+		s := len(pk.encryptedData)
+		encoded.Write([]byte{byte(s>>24), byte(s>>16), byte(s>>8), byte(s)})
+	}
+	io.Copy(w, encoded)
+	w.Write(pk.encryptedData)
 	return nil
 }
 
 func (pk *PrivateKey) serializeUnencrypted(w io.Writer) (err error) {
-	buf := bytes.NewBuffer(nil)
-	buf.Write([]byte{uint8(S2KNON)} /* no encryption */)
-	err = pk.serializePrivateKey(buf)
+	encoded := bytes.NewBuffer([]byte{uint8(S2KNON)}) /* no encryption */
+	if pk.Version == 5 {
+		encoded.Write([]byte{0x00}) /* no optional fields */
+	}
+	priv := bytes.NewBuffer(nil)
+	err = pk.serializePrivateKey(priv)
 	if err != nil {
 		return err
 	}
-	privateKeyBytes := buf.Bytes()
+	l := priv.Len() // Private material count
 	if pk.sha1Checksum {
 		h := sha1.New()
-		h.Write(privateKeyBytes)
-		sum := h.Sum(nil)
-		privateKeyBytes = append(privateKeyBytes, sum...)
+		io.Copy(h, priv)
+		priv.Write(h.Sum(nil))
```

Comment:
Sounds good to me

Reply:
`buf.Bytes()` doesn't actually allocate any memory, it just returns a slice pointing to the bytes.

Reply:
True, I was mixing up with the previous code where memory do gets allocated in `privateKeyBytes`. Anyway I'd prefer to avoid copying those private bytes too much around.

Reply:
But `privateKeyBytes := buf.Bytes()` also doesn't copy the data, it just assigns the slice which contains a pointer to the data.

Reply:
Ah, my bad then.